File: app/src/main/java/com/codeprogression/mvpb/MainView.java
package com.codeprogression.mvpb;

import android.content.Context;
import android.databinding.BaseObservable;
import android.databinding.DataBindingUtil;
import android.databinding.ObservableInt;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.view.View;
import android.widget.RelativeLayout;

import com.codeprogression.mvpb.databinding.MainViewBinding;

import javax.inject.Inject;

public class MainView extends RelativeLayout {

    @Inject
    MainPresenter presenter;
    private MainViewBinding binding;
    private ViewModel viewModel;

    public MainView(Context context, AttributeSet attrs) {
        super(context, attrs);
        if (isInEditMode()) return;
        MainActivity.component().inject(this);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (isInEditMode()) return;
        if (viewModel == null){
            viewModel = new ViewModel();
        }
        binding = DataBindingUtil.bind(this);
        binding.setViewModel(viewModel);
        binding.setListener(this);
        presenter.attach(viewModel);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        presenter.detach();
        binding.unbind();
    }

    public void add(View v) {
        // Use presenter for domain operations
        presenter.add();
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        SavedState superState = new SavedState(super.onSaveInstanceState());

        /* Save view-specific data (e.g., repeating view positions)
         * Save viewModel defaults
         */
        superState.number = viewModel.number.get();

        return superState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        if (!(state instanceof SavedState)) {
            super.onRestoreInstanceState(state);
            return;
        }

        SavedState ss = (SavedState) state;
        super.onRestoreInstanceState(ss.getSuperState());

        // update viewModel
        // update view-specific state
        viewModel = new ViewModel();
        viewModel.number.set(ss.number);
    }

    public static class ViewModel extends BaseObservable {
        public final ObservableInt number = new ObservableInt();
    }


    static class SavedState extends View.BaseSavedState {

        public int number;

        SavedState(Parcelable superState) {
            super(superState);

        }

        private SavedState(Parcel in) {
            super(in);
            number = in.readInt();
        }

        @Override
        public void writeToParcel(Parcel out, int flags) {
            super.writeToParcel(out, flags);
            out.writeInt(number);
        }

        public static final Parcelable.Creator<SavedState> CREATOR =
                new Parcelable.Creator<SavedState>() {
                    public SavedState createFromParcel(Parcel in) {
                        return new SavedState(in);
                    }

                    public SavedState[] newArray(int size) {
                        return new SavedState[size];
                    }
                };
    }

}

File: app/src/main/java/com/codeprogression/mvpb/MainView.kt
package com.codeprogression.mvpb

import android.content.Context
import android.databinding.BaseObservable
import android.databinding.DataBindingUtil
import android.databinding.ObservableInt
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import android.widget.RelativeLayout
import com.codeprogression.mvpb.databinding.MainViewBinding
import javax.inject.Inject

class MainView(context: Context, attrs: AttributeSet) : RelativeLayout(context, attrs) {

    @Inject
    lateinit internal var presenter: MainPresenter
    lateinit private var binding:  MainViewBinding
    private var viewModel: ViewModel = ViewModel()

    override fun onFinishInflate() {
        super.onFinishInflate()
        if (!isInEditMode) {
            MainActivity.component()?.inject(this)
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (isInEditMode) return
        binding = DataBindingUtil.bind(this)
        binding.viewModel = viewModel
        binding.listener = this
        presenter.attach(viewModel)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        presenter.detach()
        binding.unbind()
    }

    fun add(@Suppress("UNUSED_PARAMETER") v: View) {
        // Use presenter for domain operations
        presenter.add()
    }

    override fun onSaveInstanceState(): Parcelable {
        val superState = SavedState(super.onSaveInstanceState())

        /* Save view-specific data (e.g., repeating view positions)
         * Save viewModel defaults
         */
        superState.number = viewModel.number.get()

        return superState
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        if (state !is SavedState) {
            super.onRestoreInstanceState(state)
            return
        }

        super.onRestoreInstanceState(state.superState)

        // update viewModel
        // update view-specific state
        viewModel = ViewModel()
        viewModel.number.set(state.number)
    }

    class ViewModel : BaseObservable() {
        val number = ObservableInt()
    }

    class SavedState : View.BaseSavedState {

        var number: Int = 0

        constructor(superState: Parcelable) : super(superState) {

        }

        private constructor(`in`: Parcel) : super(`in`) {
            number = `in`.readInt()
        }

        override fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)
            out.writeInt(number)
        }

        companion object {

            val CREATOR: Parcelable.Creator<SavedState> = object : Parcelable.Creator<SavedState> {
                override fun createFromParcel(`in`: Parcel): SavedState {
                    return SavedState(`in`)
                }

                override fun newArray(size: Int): Array<SavedState?> {
                    return arrayOfNulls(size)
                }
            }
        }
    }
}
